ptrace.c

```diff
@@ -543,47 +543,76 @@ bool sm_searchregions(globals_t *vars, scan_match_type_t match_type, const userv
 
         /* load the next region */
         r = n->data;
         bytes_per_dot = r->size / NUM_DOTS;
-        bytes_at_next_dot = bytes_per_dot;
+        bytes_at_next_dot = bytes_per_dot * NUM_DOTS;
         progress_per_dot = (double)bytes_per_dot / total_scan_bytes;
 
+/* The maximum logical size is a comfortable 1MiB (increasing it does not help).
+ * The actual allocation is that plus the rounded size of the maximum possible VLT.
+ * This is needed because the last byte might be scanned as max size VLT,
+ * thus need 2^16 - 2 extra bytes after it */
+#define MAX_BUFFER_SIZE (1<<20)
+#define MAX_ALLOC_SIZE  (MAX_BUFFER_SIZE + (1<<16))
+
         /* allocate data array */
-        if ((data = malloc(r->size * sizeof(char))) == NULL) {
+        size_t alloc_size = MIN(r->size, MAX_ALLOC_SIZE);
+        if ((data = malloc(alloc_size * sizeof(char))) == NULL) {
             show_error("sorry, there was a memory allocation error.\n");
             return false;
         }
 
         /* print a progress meter so user knows we haven't crashed */
-        /* cannot use show_info here because it'll append a '\n' */
         show_user("%02u/%02u searching %#10lx - %#10lx", ++regnum,
                 vars->regions->size, (unsigned long)r->start, (unsigned long)r->start + r->size);
         fflush(stderr);
 
-        /* read region into the `data` array */
-        nread = readmemory(data, r->start, r->size);
-        if (nread == 0) {
-            show_error("reading region %02u failed.\n", regnum);
-            free(data);
-            return false;
-        }
+        /* For every offset, check if we have a match. */
+        size_t memlength = r->size;
+        size_t buffer_size = 0;
+        void *reg_pos = r->start;
+        const uint8_t *buf_pos = NULL;
+        for ( ; ; memlength--, buffer_size--, reg_pos++, buf_pos++) {
 
-        /* region has been read, tell user */
-        print_a_dot();
+            /* check if the buffer is finished (or we just started) */
+            if (UNLIKELY(buffer_size == 0)) {
 
-        /* For every offset, check if we have a match.
-         * Testing `memlength > 0` is much faster than `offset < nread` */
-        size_t memlength, offset;
-        for (memlength = nread, offset = 0; memlength > 0; memlength--, offset++) {
+                /* print a simple progress meter */
+                for ( ; memlength < bytes_at_next_dot; bytes_at_next_dot -= bytes_per_dot) {
+                    /* for user, just print a dot */
+                    print_a_dot();
+                    /* for front-end, update percentage */
+                    vars->scan_progress += progress_per_dot;
+                }
+
+                /* the whole region is finished */
+                if (memlength == 0) break;
+
+                /* stop scanning if asked to */
+                if (vars->stop_flag) break;
+
+                /* load the next buffer block */
+                size_t alloc_size = MIN(memlength, MAX_ALLOC_SIZE);
```

Comment:
Redefinition of `alloc_size` here which is misleading. Please rename this to `read_size`.

Reply:
Agreed.